R/chat.R

```diff
@@ -49,50 +55,79 @@ NULL
 #'   Briefly explain your work.
 #' ")
 new_chat_openai <- function(system_prompt = NULL,
+                            messages = NULL,
                             base_url = "https://api.openai.com/v1",
                             api_key = openai_key(),
-                            model = "gpt-4o-mini",
+                            model = NA,
                             echo = FALSE) {
   check_string(system_prompt, allow_null = TRUE)
   check_string(base_url)
   check_string(api_key)
-  check_string(model)
+  check_string(model, allow_null = TRUE, allow_na = TRUE)
   check_bool(echo)
 
-  system_prompt <- system_prompt %||%
-    "You are a helpful assistant from New Zealand who is an experienced R programmer"
+  if (is.na(model)) {
+    model <- "gpt-4o-mini"
+  }
+
+  # TODO: Sanity check individual messages?
+  if (!is.null(messages)) {
+    if (!is.list(messages) ||
+        !(is.null(names(messages)) || all(names(messages) == ""))) {
+      stop_input_type(
+        messages,
+        "an unnamed list of messages",
+        allow_null = TRUE
+      )
+    }
+  }
+
+  if (!is.null(system_prompt)) {
+    if (!is.null(messages) && length(messages) > 0) {
+      if (messages[[1]][["role"]] != "system" ||
+          !identical(messages[[1]][["content"]], system_prompt)) {
+        stop("If both `system_prompt` and `messages` are provided, `messages` must start with the `system_prompt`.")
```

Comment:
I think you could just use `check_exclusive(system_prompt, messages)` and not worry about the more complex check.

Reply:
I was just imagining users might want something like:

```r
system_prompt <- "You're a sarcastic idiot"
messages <- maybe_load_from_storage(default = NULL)

chat <- new_chat(system_prompt = system_prompt, messages = messages)
```

Which would work as long as the system prompt hadn't changed.

Your way, the last line would become:

```r
chat <- new_chat(system_prompt = if (is.null(messages)) system_prompt, messages = messages)
```

or maybe the average R programmer would do

```r
if (is.null(messages)) {
  chat <- new_chat(system_prompt = system_prompt)
} else {
  chat <- new_chat(messages = messages)
}
```

Another option would be to say the `system_prompt` is ignored if `length(messages > 0)`?

Reply:
Hmmmm, good points. We should also contemplate in the context of the claude api where the system prompt is not a message but is a separate parameter to the API?

So for openai, maybe messages should return `private$msgs[1]`, and there's a separate `system_prompt` field?

Reply:
> maybe messages should return `private$msgs[1]`

Do you mean `private$msgs[-1]`?

Reply:
Oops yes.